### PR TITLE
Add PerformanceSensitive analyzer

### PR DESCRIPTION
**Pull request by genlu**
This is a fork of [RoslynClrHeapAllocationAnalyzer](https://github.com/Microsoft/RoslynClrHeapAllocationAnalyzer), with additional support for `Roslyn.Utilities.PerformanceSensitiveAttribute`.

Majority of the commits are just porting code, consists mostly mechanical changes (converts tests from MSTest to Xunit, migrating to test harness of this repo, fixing analyzer warnings, etc.). The only meaningful code change is in https://github.com/dotnet/roslyn-analyzers/commit/0ea85227364ccbaa3c5f534af97a9b8e533901dc, which make those analyzer understand `PerformanceSensitiveAttribute`. Right now the support is minimal, i.e. analysis will be triggered only for code block decorated with this attribute, but actual values of the parameters are ignored.

TODO:

- [x] LOC
- [ ] Make analyzers respect parameters of `PerformanceSensitiveAttribute`
- [ ] editorconfig support to change the analysis scope (code block with `PerformanceSensitiveAttribute` only vs. everything)

### CODE REVIEW

**Comment by genlu** [comment]
FYI @jinujoseph

**Comment by sharwell** [comment]
> migrating to test harness of this repo

Oh no 😂  :trollface:

File: src/PerformanceSensitive.Analyzers/Core/AbstractAllocationAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                syntaxNodeContext =>
+                {
+                    AnalyzeNode(syntaxNodeContext, in info);
+                },
+                Expressions);
+        }
+
+        protected sealed class AttributeChecker
+        {
+            private INamedTypeSymbol PerfSensitiveAttributeSymbol { get; }
```
Comment (sharwell) [inline src/PerformanceSensitive.Analyzers/Core/AbstractAllocationAnalyzer.cs]: 💡 Spell out `Performance`

**Comment by sharwell** [comment]
@genlu I'm going to push a commit onto this branch to convert the tests forward one more time to Microsoft.CodeAnalysis.Testing.

**Comment by mavasani** [comment]
> I think we should take this one as-is and move the remaining items to follow-up issues.

I strongly disagree. I understand the desire to make progress on PerformanceSensitiveAttribute work, but porting such a huge set of analyzers, which do not follow the repo standards and implementation in multiple ways, and would require a significant rewrite is certainly _not_ the way to go. Let us discuss this at a design meeting before dumping such a huge pile of code in the repo and then iteratively rewriting pieces.

**Comment by jmarolf** [comment]
Selfishly I would like this analyzer to be re-written to use `IOperation`.  This is the classic case for why `IOperation` was needed and validating that would be great! It would also help with some of the correctness issues in the current implementation.

**Comment by mavasani** [comment]
@sharwell mentioned offline that this PR is needed to go in asap to unblock consuming these analyzer in Roslyn.sln for PerformanceSensitiveAttribute related work. If so, we have following options:

1. Take this PR as-is as a short term solution, with known performance and/or implementation divergence from rest of this repo. Someone should be on point to help address the long term concerns and/or cleanup work to bring these analyzers in line with this repo. My primary fear is these will just be tracked as issues that never get looked at again.
2. Clone the code from this PR into a private repo or DotNetAnalyzers and unblock Sam's work. Once we have figured out the long term logistics for these analyzers, we should selectively port pieces into this repo ensuring they meet those long term goals and repo standards. At that point we can delete the implementation from DotNetAnalyzers and eventually even enable Mukul to delete his repo.

My primary point is let us make conscious choices here given we are forking analyzers. Mukul really wants us to take ownership of these analyzers and delete his repo, and a single, performant implementation using IOperation that serves everyone's needs and helps us kill all other cloned implementations should be our long term goal. I can remove my block if @jinujoseph @vatsalyaagrawal feel approach 1. is the way to go here.

**Comment by mavasani** [comment]
@jinujoseph @vatsalyaagrawal I also feel we need to make more long term decisions on how to handle temporary/WIP analyzer implementations in this repo. For example, https://github.com/dotnet/roslyn-analyzers/pull/1111 ports the Naming service and first FxCop naming rule, which unblocks porting [Naming rules](https://docs.microsoft.com/en-us/visualstudio/code-quality/naming-warnings?view=vs-2017) from legacy FxCop to analyzers in this repo. The PR has been blocked for more than a year as the implementation does not follow the standards of this repo and has lot of cloned code from other repos/legacy FxCop. Given that Naming rules are the primary category of rules that have not been ported from legacy FxCop and have multiple customer requests from those who have attempted to migrate to FxCop analyzers, they are actually much higher priority for us for FxCop migration. I feel the approach we take for the needs here should be followed on that PR too and should be used for all such WIP contributions in future.

**Comment by sharwell** [comment]
> My primary fear is these will just be tracked as issues that never get looked at again.

Most of the IOperation ports are getting tracked by me as work to reduce `GetSymbolInfo` costs during builds. Right now these issues are spread across many different repositories implementing analyzers based on the relative measured impact on builds. The dominant impact in this space right now is vs-threading analyzers, but it would get reprioritized if something else started to show up. I'm not especially concerned by PerformanceSensitive analyzer primarily because the impact is bounded by the use of PerformanceSensitiveAttribute, which by design is not frequently used.

**Comment by mavasani** [comment]
> I'm not especially concerned by PerformanceSensitive analyzer primarily because the impact is bounded by the use of PerformanceSensitiveAttribute, which by design is not frequently used.

That is true for the use case that you are driving. We have committed to Mukul to take up ownership of these analyzers, in context of them operating on a whole code base. As soon as we do that or add an editorconfig option for it like @genlu mentioned in the PR description, we are signing up to ship an implementation that has known performance drawback.

Again, I am fine with this PR as a short term, WIP implementation to unblock your use case. However, we need to clarify this is a viable approach for all such current and future scenarios. This also implies we should take https://github.com/dotnet/roslyn-analyzers/pull/1111 as-is and then work towards follow-up items to cleanup/improve the code added there.

**Comment by genlu** [comment]
I have add localization and renamed the package to Microsoft.PerformanceSensitive.Analyzers per our offline discussion. Please take another look before we merge.

@mavasani @sharwell

**Comment by genlu** [comment]
I have filed #2120 to track IOperation conversion work, besides what we a have already discussed, we will need that for the analyzer to work in VB.

**Comment by mavasani** [comment]
Tagging @mjsabby for reviewing the PR, he wanted to give inputs on the attribute as he also has a PR contributing similar functionality: https://github.com/Microsoft/RoslynClrHeapAllocationAnalyzer/pull/61

**Comment by mjsabby** [comment]
I support moving to IOperations and the Roslyn team taking over the original analyzer.

From one of our issues, I dug this up. This is what I really wanted in a PerformanceContractAttribute or as you've called it PerformanceSensitiveAttribute.

The idea is that there will be methods where we can go without any allocations, and I'd like to break the build or warn. Furthermore, certain types of allocations are OK, but certain kinds are not. So this would capture that.

I haven't looked at the PR yet, but would like that to be included.

```csharp
    using System;

    [AttributeUsage(AttributeTargets.Constructor | AttributeTargets.Method | AttributeTargets.Property | AttributeTargets.Field, AllowMultiple = true, Inherited = false)]
    public class PerformanceContractAttribute : Attribute
    {
        public DiagnosticType Diagnostic { get; set; }

        public AllocationKind AllowedAllocations { get; set; }
    }

    public enum DiagnosticType
    {
        Warning,
        Error,
        Notify
    }

    [Flags]
    public enum AllocationKind
    {
        None,
        Boxing,
        ImplicitArrayAllocation,
        Captures,
        Explicit,
        Enumerator,
        Delegate,
        All = Boxing | ImplicitArrayAllocation | Captures | Explicit | Enumerator | Delegate // useful when the analyzer setting is set to complain about everything and you still need an escape hatch
    }

    public class Foo
    {
        [PerformanceContract(AllowedAllocations = AllocationKind.Explicit | AllocationKind.Delegate)]
        public void RegularMethod()
        {
        }


        [PerformanceContract(AllowedAllocations = AllocationKind.None, Diagnostic = DiagnosticType.Error)]
        public void SuperHotMethod()
        {
        }

        [PerformanceContract(AllowedAllocations = AllocationKind.All)]
        public void InitializationOrStartupMethod()
        {
        }
    }
```

**Comment by genlu** [comment]
What you proposed is similar to the properties what we have today, but it allows more flexibility in terms of controlling warning level and suppress warnings. Will need to discuss with @sharwell on if/how to cooperate those into PerformanceSensitiveAttribute. 

Anyway, it's something need to be done in a follow up PR. We will keep you in the loop for any change in this area going forward. Meanwhile, any objection on merging this?

**Comment by mjsabby** [comment]
Yes that’s what I also noticed. I think if we can support this that’d be very useful for the direction I was taking.

Overall our goals aligned, and if we can have  the flexibility I proposed I’ll be a happy camper.

No objections

**Pull request merged**